File: backend-php/src/http.php
<?php

declare(strict_types=1);

use Psr\Http\Message\ResponseInterface as Response;
use Psr\Http\Message\ServerRequestInterface as Request;
use Slim\Factory\AppFactory;

const TRINO_UNAVAILABLE = [
    'error' => 'Trinoに接続できませんでした。稼働状況を確認して、もう一度お試しください。',
    'code' => 'trino_unavailable',
];

function normalize_filters(array $args): array
{
    return [
        'date' => trim((string) ($args['date'] ?? '')),
        'time_from' => trim((string) ($args['time_from'] ?? '')),
        'time_to' => trim((string) ($args['time_to'] ?? '')),
        'log_type' => trim((string) ($args['log_type'] ?? '')),
        'host' => trim((string) ($args['host'] ?? '')),
        'program' => trim((string) ($args['program'] ?? '')),
        'message' => trim((string) ($args['message'] ?? '')),
        'page' => positive_int($args['page'] ?? 1, 1),
        'size' => min(positive_int($args['size'] ?? 25, 25), 100),
        'skip_total' => in_array(strtolower((string) ($args['skip_total'] ?? '')), ['1', 'true'], true),
    ];
}

function positive_int(mixed $value, int $fallback): int
{
    $parsed = filter_var($value, FILTER_VALIDATE_INT);
    return is_int($parsed) && $parsed > 0 ? $parsed : $fallback;
}

function filters_from_request(Request $request): array
{
    $contentType = $request->getHeaderLine('Content-Type');
    if (str_contains($contentType, 'application/json')) {
        $payload = $request->getParsedBody();
        if ($payload === null) {
            $payload = json_decode((string) $request->getBody(), true);
        }
        return normalize_filters(is_array($payload) ? $payload : []);
    }

    if ($request->getMethod() === 'POST') {
        $payload = $request->getParsedBody();
        return normalize_filters(is_array($payload) ? $payload : []);
    }

    return normalize_filters($request->getQueryParams());
}

function json_response(Response $response, array $payload, int $status = 200): Response
{
    $response->getBody()->write(json_encode($payload, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES));
    return $response
        ->withHeader('Content-Type', 'application/json; charset=utf-8')
        ->withStatus($status);
}

function create_app(): \Slim\App
{
    $app = AppFactory::create();
    $app->addBodyParsingMiddleware();
    $app->addRoutingMiddleware();

    $config = app_config();

    $app->get('/', function (Request $request, Response $response) use ($config): Response {
        return json_response($response, [
            'service' => 'php-trino-backend',
            'endpoints' => ['/health', '/api/options', '/api/logs', '/api/summary'],
        ]);
    });

    $app->get('/health', function (Request $request, Response $response) use ($config): Response {
        return json_response($response, [
            'ok' => trino_ping($config),
            'trino_url' => $config['trino_url'],
            'catalog' => $config['trino_catalog'],
            'schema' => $config['trino_schema'],
        ]);
    });

    $app->get('/api/options', function (Request $request, Response $response): Response {
        return json_response($response, ['log_types' => LOG_TYPES]);
    });

    $app->map(['GET', 'POST'], '/api/logs', function (Request $request, Response $response) use ($config): Response {
        $filters = filters_from_request($request);
        try {
            $result = search_logs_page($filters, $config);
            return json_response($response, ['filters' => $filters, ...$result]);
        } catch (Throwable $error) {
            error_log('Trino log search failed: ' . $error);
            return json_response($response, TRINO_UNAVAILABLE, 502);
        }
    });

    $app->get('/api/summary', function (Request $request, Response $response) use ($config): Response {
        try {
            $date = trim((string) ($request->getQueryParams()['date'] ?? ''));
            return json_response($response, get_log_total($date, $config));
        } catch (Throwable $error) {
            error_log('Trino log summary failed: ' . $error);
            return json_response($response, TRINO_UNAVAILABLE, 502);
        }
    });

    $app->addErrorMiddleware(true, true, true);

    return $app;
}

File: backend-go/http.go
package main

import (
	"encoding/json"
	"errors"
	"io"
	"log"
	"math"
	"net/http"
	"strconv"
	"strings"
)

type App struct {
	client TrinoExecutor
}

var trinoUnavailable = map[string]any{
	"error": "Trinoに接続できませんでした。稼働状況を確認して、もう一度お試しください。",
	"code":  "trino_unavailable",
}

func NewApp(client TrinoExecutor) (*App, error) {
	return &App{client: client}, nil
}

func (a *App) routes() http.Handler {
	mux := http.NewServeMux()
	mux.HandleFunc("/", a.index)
	mux.HandleFunc("/health", a.health)
	mux.HandleFunc("/api/options", a.apiOptions)
	mux.HandleFunc("/api/logs", a.apiSearchLogs)
	return mux
}

func (a *App) index(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	writeJSON(w, map[string]any{
		"service":   "go-trino-backend",
		"endpoints": []string{"/health", "/api/options", "/api/logs"},
	})
}

func (a *App) apiOptions(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	writeJSON(w, map[string]any{"log_types": logTypes})
}

func (a *App) health(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	writeJSON(w, map[string]any{
		"ok":        a.client.Ping(r.Context()),
		"trino_url": trinoURL,
		"catalog":   trinoCatalog,
		"schema":    trinoSchema,
	})
}

func (a *App) apiSearchLogs(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet && r.Method != http.MethodPost {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	filters, err := filtersFromRequest(r)
	if err != nil {
		writeJSONStatus(w, http.StatusBadRequest, map[string]any{"error": err.Error()})
		return
	}
	logs, total, err := searchLogsPage(r.Context(), a.client, filters)
	if err != nil {
		log.Printf("Trino log search failed: %v", err)
		writeJSONStatus(w, http.StatusBadGateway, trinoUnavailable)
		return
	}

	writeJSON(w, map[string]any{
		"filters":     filters,
		"count":       len(logs),
		"total":       total,
		"page":        filters.Page,
		"size":        filters.Size,
		"total_pages": int(math.Max(1, math.Ceil(float64(total)/float64(filters.Size)))),
		"logs":        logs,
	})
}

func filtersFromRequest(r *http.Request) (Filters, error) {
	if strings.HasPrefix(r.Header.Get("Content-Type"), "application/json") {
		var filters Filters
		if r.Body != nil {
			if err := json.NewDecoder(r.Body).Decode(&filters); err != nil && !errors.Is(err, io.EOF) {
				return Filters{}, err
			}
		}
		return normalizeFilters(filters), nil
	}
	if r.Method == http.MethodPost {
		if err := r.ParseForm(); err != nil {
			return Filters{}, err
		}
		return filtersFromValues(r.PostForm), nil
	}
	return filtersFromValues(r.URL.Query()), nil
}

func filtersFromValues(values map[string][]string) Filters {
	get := func(key string) string {
		if len(values[key]) == 0 {
			return ""
		}
		return values[key][0]
	}
	return normalizeFilters(Filters{
		Date: get("date"), TimeFrom: get("time_from"), TimeTo: get("time_to"),
		LogType: get("log_type"), Host: get("host"), Program: get("program"), Message: get("message"),
		Page: parsePositiveInt(get("page"), 1), Size: parsePositiveInt(get("size"), 25),
	})
}

func normalizeFilters(filters Filters) Filters {
	size := filters.Size
	if size <= 0 {
		size = 25
	}
	if size > 100 {
		size = 100
	}
	page := filters.Page
	if page <= 0 {
		page = 1
	}
	return Filters{
		Date: strings.TrimSpace(filters.Date), TimeFrom: strings.TrimSpace(filters.TimeFrom),
		TimeTo: strings.TrimSpace(filters.TimeTo), LogType: strings.TrimSpace(filters.LogType),
		Host: strings.TrimSpace(filters.Host), Program: strings.TrimSpace(filters.Program),
		Message: strings.TrimSpace(filters.Message), Page: page, Size: size,
	}
}

func parsePositiveInt(value string, fallback int) int {
	number, err := strconv.Atoi(value)
	if err != nil || number <= 0 {
		return fallback
	}
	return number
}

func writeJSON(w http.ResponseWriter, value any) {
	writeJSONStatus(w, http.StatusOK, value)
}

func writeJSONStatus(w http.ResponseWriter, status int, value any) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	encoder := json.NewEncoder(w)
	encoder.SetEscapeHTML(false)
	_ = encoder.Encode(value)
}
